src/coreclr/tools/Common/TypeSystem/IL/KnownReferences.cs

```diff
@@ -0,0 +1,264 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using ILCompiler;
+using Internal.TypeSystem;
+using Internal.TypeSystem.Ecma;
+
+namespace Internal.IL
+{
+    public enum KnownILStubReference
+    {
+        Exception,
+
+        ExecutionAndSyncBlockStore,
+        ExecutionAndSyncBlockStore_Push,
+        ExecutionAndSyncBlockStore_Pop,
+
+        AsyncHelpers,
+
+        AsyncHelpers_AsyncCallContinuation,
+        AsyncHelpers_TransparentAwait,
+        AsyncHelpers_CompletedTask,
+        AsyncHelpers_CompletedTaskResult,
+
+        AsyncHelpers_TaskFromException,
+        AsyncHelpers_TaskFromException_1,
+        AsyncHelpers_ValueTaskFromException,
+        AsyncHelpers_ValueTaskFromException_1,
+
+        AsyncHelpers_FinalizeTaskReturningThunk,
+        AsyncHelpers_FinalizeTaskReturningThunk_1,
+        AsyncHelpers_FinalizeValueTaskReturningThunk,
+        AsyncHelpers_FinalizeValueTaskReturningThunk_1,
+
+        Task,
+        Task_1,
+        ValueTask,
+        ValueTask_1,
+
+        Task_FromResult_1,
+        ValueTask_FromResult_1,
+
+        Task_get_CompletedTask,
+        ValueTask_get_CompletedTask,
+
+        Task_get_IsCompleted,
+        ValueTask_get_IsCompleted,
+        ValueTask_1_get_IsCompleted,
+
+        ValueTask_ThrowIfCompletedUnsuccessfully,
+        ValueTask_1_get_Result,
+
+        ValueTask_AsTaskOrNotifier,
+        ValueTask_1_AsTaskOrNotifier,
+
+        AsyncHelperCount,
+    }
+
+    public class KnownILStubReferences
+    {
+        /// <summary>
+        /// Extracts all method and type references from the IL of a method.
```

Comment:
If we are doing scanning to be "correct by construction", do we still need the `KnownILStubReference` enum?

Reply:
I think we should be able to get rid of the enum and leave AsyncThunk.cs as it was before this PR, without the enum indirections.

If we're concerned that someone might modify AsyncThunks.cs and add references to other CoreLib internals that we did not approve as "non-versionable from R2R perspective", we could add a check to `ExtractReferencesFromIL` that checks the reference against an approved whitelist. But I don't think we need an enum for the whitelist, that's just extra indirections. This can just be a big switch block over `method.Name` with some `case "Foo" when ...`.

Reply:
> This can just be a big switch block over method.Name with some case "Foo" when ....

We could also consider adding an internal attribute to CoreLib - similar to NonVersionableAttribute - that would mark APIs that are okay to generate references to from R2R IL stubs. Then the check for valid APIs could just look for the attribute, no need to duplicate things for the whitelist.